### PR TITLE
Site icon: Fix site icon styling to display non-square site icons within a square button

### LINKED ISSUE #37564: Site Icon: Inconsistent hover styling
### Description

#29888 added hover styling to the site button in the Header, including an outline that appears around the Site Icon when hovered. #33132 later added a zooming animation to the button's hover state.

Currently this means:
* If no Site Icon exists, the default icon renders and receives the outline but no animation when hovered
* If a Site Icon exists, the animation runs and 'zooms in' the icon. For a square icon, this causes it to perfectly fill the container, but for a non-square icon the outline is still partially visible behind the button. See screenshots below.

This could be addressed by removing the outline and ensuring the default icon also gets the animation. Alternatively the outline should at least be removed when a Site Icon has been set, to avoid the issue with non-square icons.

### Step-by-step reproduction instructions

See the videos below

1. Navigate to the Customizer (`/wp-admin/customize.php`) -> Site Identity and upload a new Site Icon. Use an image that is not square, and select "Skip Cropping" when prompted to keep the original image. Save changes.
2. Open a post in the block editor.
3. Hover over the button in the top-left of the Header to see the animation and visible outline.
4. Now go back to the Customizer and try uploading a new Site Icon, but make sure to crop it this time. Publish.
5. Open the block editor and hover over the site button; observe that the icon fills the button and no outline is visible.
6. Go back to the Customizer, remove the site Icon, and Publish.
7. Open the block editor and observe that the default icon renders. Hover over it and see that it does not zoom in, and that the outline renders.

### Screenshots, screen recording, code snippet

Default Icon (when no Site Icon has been set):

https://user-images.githubusercontent.com/63313398/146990859-022d4ffb-4d7a-4340-a22d-50c0a4fa3cfc.mp4

Square Site Icon (upload a square image, or use the cropping tool in the Customizer):

https://user-images.githubusercontent.com/63313398/146990916-db9155da-fb81-4440-b13b-d9e9ffa55450.mp4

Non-square Site Icon (upload a non-square image and select `Skip Cropping`):

https://user-images.githubusercontent.com/63313398/146990968-7da529ad-7268-4a1a-a562-4a9689085f6f.mp4

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by andrewserong**
Fixes #37564

## Description

Following on from the site icon feature introduced in #35892 this PR fixes the styling for the site icon in the W button so that it displays non-square site icons within a square button. The fix does the following:

* Switch the W logo button image to use a `div` instead of an `img` tag, and set the image as a background image with background size of `cover`.
* Roll out styling to both the site and post editors

Note: I needed to add the `-1px` on the top margin to get the image to be centred correctly within the border as below:

image

I _think_ this is because the container is `61px` high instead of `60px` to deal with the grey border, so getting the position exact along with the scale animation is slightly tricky. Let me know if anyone thinks this needs tweaking further!

## How has this been tested?

To quickly add a non-square site icon:

1. Using the Site Logo block, set the block to `Use as site icon` in the sidebar inspector controls.
2. Upload a non-square site logo and update the post/page to save the Icon entity.
3. Load the post editor and test that the non-square icon is displayed as a square in the top left and that when hovered it correctly fills the whole square of the button area
4. Load the site editor and test that the non-square icon is displayed as a square in the top left and that when hovered it correctly fills the whole square of the button area, also when you click on it, the highlighted border should be around the outside of the icon as in the screenshot above
5. Try the above with a few differently shaped logos, including square, tall, and narrow icons

Also, I checked that the styling when no icon is set appears to be unaffected (to test this go to: `/wp-admin/customize.php?autofocus[section]=title_tagline` and clear the existing site icon, then reload the post and site editors)

## Screenshots 

| Before | After | 
| --- | --- |
| Kapture 2021-12-22 at 12 00 27 | Kapture 2021-12-22 at 12 05 58 |

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by andrewserong** [comment]
James and Joen, I've added you both as reviewers if you get a chance, since you've given feedback to or contributed to the site icon recently 🙂

**Comment by glendaviesnz** [comment]
This tested well for me with wide, tall, and square images in both post and site editor contexts. Technically this looks like a good solution to me, but will hold off giving it a ✅ until it has had design feedback.

**Comment by jasmussen** [comment]
My goodness, I hadn't even noticed that the hover style for a customized logo had changed! Here are GIFs to help the conversation along. In trunk, a landscape image remains landscape, just scaled down. It has a focus style, and it zooms way in on hover:
trunk

This branch is the same, only the landscape image is soft-cropped to be square, regardless of its original aspect ratio:

this branch

Just looking at the before and after, this is better, the image should be square. 👍 👍 — I personally think the hover style might be a bit jumpy, but it's fine, and it's separate.

In principle the code looks good, however it seems like you should be able to accomplish the square crop by combining the same width and height with `object-fit: cover;`. (I was just looking into [soft-cropping](https://codepen.io/joen/pen/mdBMwQR), in case you'd like to dive into `aspect-ratio` as well. I don't necessarily think it's problematic to move it to use a div with a background instead of an image, but in case you wanted to make the changeset smaller, theoretically adding the height and object-fit might do it.

File: packages/edit-post/src/components/header/fullscreen-mode-close/style.scss
Comment (jasmussen) [inline packages/edit-post/src/components/header/fullscreen-mode-close/style.scss]: Excellent eye. Interestingly, when you're just toggling off that minus 1 margin in the inspector, it makes the site icon look vertically uncentered when it's on:
margin shift

However when you zoom in and look at the details, you can see that it has to have that negative margin in order to horizontally align with the other buttons in the toolbar. With negative margin:
with minus 1 px

Without negative margin:

without minus 1 px

This _is_, per your instinct, related to the dark square being 61px tall. That math comes $header-height + $border-width, so that the black area covers the gray border below the top-bar.

So which do we keep? I don't have a strong opinion, but I appreciate your attention to detail, so probably do keep it. But I'd add a small comment above, something like `// Compensate for the top-bar border.`, and then change `1px` to `$border-width` instead.

File: packages/edit-site/src/components/navigation-sidebar/navigation-toggle/style.scss
Comment (jasmussen) [inline packages/edit-site/src/components/navigation-sidebar/navigation-toggle/style.scss]: Same as above, could use a comment and a variable.

**Review by jasmussen — APPROVED** [review]
I noted in my testing that you could have probably accomplished the same square icon with `object-fit: cover;`. I would encourage a quick code sanity check to see if that's actually better, as that would afford the alt text to remain. Whether that's actually better for a screen-reader, to read a button with an image inside, or to just read a button, I can't say for sure.

But just as a before/after enhancement, this one is good to go.

File: packages/edit-post/src/components/header/fullscreen-mode-close/style.scss
Comment (jameskoster) [reply packages/edit-post/src/components/header/fullscreen-mode-close/style.scss]: Might it be neater to make the black square 60px tall, then add a shadow to hide the grey border? Then we don't need the awkward negative margin.

Comment (andrewserong) [reply packages/edit-post/src/components/header/fullscreen-mode-close/style.scss]: Thanks for the detailed approach here! I had a go at switching to 60px tall and using a shadow to hide the grey border. Unfortunately the styling in the site editor then wound up being a bit unreliable due to the `z-index` stacking order. With `60px` and box-shadow, from time to time the shadow would pop out / a white line would be visible:

Kapture 2021-12-23 at 11 32 51

I've gone with adding the inline comment and using `$border-width` instead of the hard-coded `1px`, and we can always look at switching to the `60px` approach in a follow-up if we'd like to tweak further.

File: packages/edit-site/src/components/navigation-sidebar/navigation-toggle/style.scss
Comment (andrewserong) [reply packages/edit-site/src/components/navigation-sidebar/navigation-toggle/style.scss]: Switched to using the variable, and added a comment.

**Comment by andrewserong** [comment]
Thanks for the review @jasmussen!

> I noted in my testing that you could have probably accomplished the same square icon with object-fit: cover;. I would encourage a quick code sanity check to see if that's actually better, as that would afford the alt text to remain.

Yes, I keep forgetting that we can use `object-fit` now that we no longer need to support IE11 😀, that works well and means we only need a CSS change for this PR instead of also updating the JS. I've moved over to that approach and re-tested and it still appears to be working well. Once the tests pass, I'll merge this in.

I liked the idea of using the `60px` height, but I ran into some issues attempting to implement it (https://github.com/WordPress/gutenberg/pull/37570#discussion_r774251096). Since I'm about to go AFK, I think it's better to get this fix in as-is, and we can always look at tweaking the styling in a follow-up (I'm happy to take a look next year if no-one beats me to it).

**Pull request merged**
